src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3139,10 +3139,24 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           val initElems = scope.elems
           // SI-5877 The decls of a package include decls of the package object. But we don't want to add
           //         the corresponding synthetics to the package class, only to the package object class.
-          def shouldAdd(sym: Symbol) =
-            inBlock || !context.isInPackageObject(sym, context.owner)
+          // SI-6734 Locality test below is meaningless if we're not even in the correct tree.
+          def shouldAdd(sym: Symbol): Boolean = {
+            def shouldAddAsModule: Boolean =
+              sym.moduleClass.attachments.get[ClassForCaseCompanionAttachment] match {
+                case Some(att) =>
+                  val cdef = att.caseClass
+                  stats.exists {
```

Comment:
This code still looks strange to me. The attachment **is** a `ClassDef`, but the code does a linear search for a(nother?) `ClassDef` with the same symbol. Isn't it the same `ClassDef`? If not, why do they have the same symbol? In either case it might be good to document it.

This may be perfectly ok, but it's still not clear to me why it only treats symbols that are currently compiled (that have trees around).